### PR TITLE
Permit duplicate param logging

### LINKED ISSUE #523: Fix parameter logging of duplicates
Issue #43 was not completely implemented. For compatibility with MLFlow tracking server, we need:
1) 400 error when batch contains same key/different value parameter (currently returns 500)
2) 200 for same key/same value needs to work across batches, not just within batches.

### PR DESCRIPTION
**Pull request by suprjinx**
Previous iteration on this story only permitted duplicate parameterss within a batch. The changes here rely on the database to detect  permitted duplication.

Additionally, the error status code for parameters conflicts was changed from 500 ISE to 400 Bad Request.
Fixes #523

### CODE REVIEW

File: pkg/api/mlflow/dao/repositories/param.go [hunk 1]
```diff
@@ -32,48 +43,65 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
-	var keys []string
-	paramMap := map[string]models.Param{}
-	for _, param := range params {
-		key := fmt.Sprintf("%v-%v-%v", param.RunID, param.Key, param.Value)
-		keys = append(keys, key)
-		paramMap[key] = param
+// findConflictingParams checks if there are conflicting values for the input params. If a key does not
+// yet exist in the db, or if the same key and value already exist for the run, it is not a conflict.
+// If the key already exists for the run but with a different value, it is a conflict. Conflicting keys are returned.
+func (r ParamRepository) findConflictingParams(ctx context.Context, params []models.Param) ([]string, error) {
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/dao/repositories/param.go:72]: this whole function could be changed by a single SQL query in this spirit:
```sql
WITH new(key, value) AS (VALUES (?, ?), (?, ?))
SELECT current.key, current.value, new.value
FROM params AS current
INNER JOIN new ON run_uuid = ? AND new.key = current.key
WHERE new.value != current.value;
```

`VALUES` can be built as a list of  `n` `"(?, ?)"` joined by `", "`, where `n` is the amount of params we tried to insert.

The SQL parameters that need to be passed are the list keys and values we tried to insert, and then the run ID.

The DB engine will then return to you all you need to log the error nicely to the user.

Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:72]: I can change to this, but I'd suggest a simple query like we have currently plus Go-layer manipulation is easier to grok

File: pkg/api/mlflow/dao/repositories/param.go [hunk 2]
```diff
@@ -32,48 +43,65 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
 // CreateBatch creates []models.Param entities in batch.
 func (r ParamRepository) CreateBatch(ctx context.Context, batchSize int, params []models.Param) error {
-	// try to create params in batch; error condition requires special handling
-	// to allow certain duplicates
-	if err := r.db.CreateInBatches(params, batchSize).Error; err != nil {
-		// remove duplicate rows and try again
-		dedupedParams, errRemovingMatches := r.removeExactMatches(ctx, params)
-		if errRemovingMatches != nil {
-			return eris.Wrap(errRemovingMatches, "error removing duplicates in batch")
+	tx := r.db.Clauses(clause.OnConflict{
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:46]: done

File: pkg/api/mlflow/dao/repositories/param.go [hunk 3]
```diff
@@ -32,48 +43,65 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
-		paramMap[key] = param
+// findConflictingParams checks if there are conflicting values for the input params. If a key does not
+// yet exist in the db, or if the same key and value already exist for the run, it is not a conflict.
+// If the key already exists for the run but with a different value, it is a conflict. Conflicting keys are returned.
+func (r ParamRepository) findConflictingParams(ctx context.Context, params []models.Param) ([]string, error) {
+	dbParams := []models.Param{}
+	paramKeysInError := []string{}
+	if err := r.db.WithContext(ctx).
+		Model(&models.Param{}).
+		Where("run_uuid = ?", params[0].RunID).
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:77]: this assumption is removed in rewriting the query to find duplicate

File: pkg/api/mlflow/dao/repositories/param.go [hunk 4]
```diff
@@ -32,48 +43,65 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
-	for _, foundKey := range foundKeys {
-		delete(paramMap, foundKey)
+// collectKeys collects the keys from the params.
+func (r ParamRepository) collectKeys(params []models.Param) []string {
+	keys := make([]string, len(params))
+	for i, param := range params {
+		keys[i] = param.Key
 	}
+	return keys
+}
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:98]: removed receiver

File: pkg/api/mlflow/dao/repositories/param.go [hunk 5]
```diff
@@ -43,65 +44,60 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
+	params []models.Param,
+) ([]map[string]string, error) {
+	var paramsInError []map[string]string
 	if err := r.db.WithContext(ctx).
-		Model(&models.Param{}).
-		Where("run_uuid = ?", params[0].RunID).
-		Where("key IN ?", r.collectKeys(params)).
-		Find(&dbParams).Error; err != nil {
+		Raw(fmt.Sprintf(`
+		    WITH new(key, value, run_uuid) AS (VALUES %s)
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/dao/repositories/param.go:84]: this is vulnerable to SQL injection
as I explained in https://github.com/G-Research/fasttrackml/pull/550#discussion_r1382611435, please build the values as a list of `n` `"(?, ?)"` strings joined by `", "`, and then pass the actual values as parameters to the query

File: pkg/api/mlflow/dao/repositories/param.go [hunk 6]
```diff
@@ -43,65 +44,60 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
-func (r ParamRepository) findConflictingParams(ctx context.Context, params []models.Param) ([]string, error) {
-	dbParams := []models.Param{}
-	paramKeysInError := []string{}
+// If the key already exists for the run but with a different value, it is a conflict. Conflicting keys
+// are returned.
+func (r ParamRepository) findConflictingParams(
+	ctx context.Context,
+	params []models.Param,
+) ([]map[string]string, error) {
+	var paramsInError []map[string]string
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/dao/repositories/param.go:81]: create an actual struct type instead

Comment (jgiannuzzi) [reply pkg/api/mlflow/dao/repositories/param.go:81]: e.g.
```golang
type errorParam struct {
	RunID string
	Key string
	OldValue string
	NewValue string
}
```

File: pkg/api/mlflow/dao/repositories/param.go [hunk 7]
```diff
@@ -43,65 +44,60 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
 	if err := r.db.WithContext(ctx).
-		Model(&models.Param{}).
-		Where("run_uuid = ?", params[0].RunID).
-		Where("key IN ?", r.collectKeys(params)).
-		Find(&dbParams).Error; err != nil {
+		Raw(fmt.Sprintf(`
+		    WITH new(key, value, run_uuid) AS (VALUES %s)
+		    SELECT current.key as key, current.value as old_value, new.value as new_value
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/dao/repositories/param.go:85]: you also need to select the run ID if we want to be run agnostic

Comment (jgiannuzzi) [reply pkg/api/mlflow/dao/repositories/param.go:85]: alternatively, change the ParamRepositoryProvider interface as discussed before

File: pkg/api/mlflow/dao/repositories/param.go [hunk 8]
```diff
@@ -43,65 +44,60 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
+		    FROM params AS current
+		    INNER JOIN new ON new.run_uuid = current.run_uuid AND new.key = current.key
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/dao/repositories/param.go:87]: if we go for being run-agnostic, you can write this as `INNER JOIN new USING(run_uuid, key)`

File: pkg/api/mlflow/dao/repositories/param.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply pkg/api/mlflow/dao/repositories/param.go]: could you please add a test for this?

File: pkg/api/mlflow/dao/repositories/param.go (hunk 8 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go]: done

File: pkg/api/mlflow/dao/repositories/param.go (hunk 7 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go]: including run id for each param

File: pkg/api/mlflow/dao/repositories/param.go (hunk 6 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go]: will do

File: pkg/api/mlflow/dao/repositories/param.go (hunk 5 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go]: done

File: pkg/api/mlflow/dao/repositories/param.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go]: added test in integration tests (mlflow/run/log_batch_test.go)

File: pkg/api/mlflow/dao/repositories/param.go [hunk 9]
```diff
@@ -72,32 +85,41 @@ func (r ParamRepository) CreateBatch(ctx context.Context, batchSize int, params
...
-		    INNER JOIN new ON new.run_uuid = current.run_uuid AND new.key = current.key
-		    WHERE new.value != current.value
-                `, prepareSqlValues(params))).
-		Scan(&paramsInError).Error; err != nil {
-		return nil, eris.New("error fetching params from db")
+// If the key already exists for the run but with a different value, it is a conflict. Conflicts are returned.
+func findConflictingParams(tx *gorm.DB, params []models.Param) ([]paramConflict, error) {
+	var conflicts []paramConflict
+	if err := tx.Raw(
+		`WITH new(key, value, run_uuid) AS (VALUES `+makeSqlPlaceholders(params)+`)
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/dao/repositories/param.go:92]: could you please use `fmt.Sprintf` here to improve readability?

File: tests/integration/golang/mlflow/run/log_batch_test.go
Comment (jgiannuzzi) [inline tests/integration/golang/mlflow/run/log_batch_test.go:1]: we're missing a test of duplicate key with different value in a _separate batch_

File: pkg/api/mlflow/dao/repositories/param.go (hunk 9 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:92]: i started with that but it seemed like this was clearer -- since we use a values array for sql parameter replacement, it seemed weird to have first one value replacement for the template string (sql), then another

Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go]: looks better w/ separate statement to make the sql template, please take a look

File: tests/integration/golang/mlflow/run/log_batch_test.go
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_batch_test.go:1]: added a setup to to put "previous batch" param in db

File: pkg/api/mlflow/dao/repositories/param.go [hunk 10]
```diff
@@ -32,48 +57,68 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
 // CreateBatch creates []models.Param entities in batch.
 func (r ParamRepository) CreateBatch(ctx context.Context, batchSize int, params []models.Param) error {
-	// try to create params in batch; error condition requires special handling
-	// to allow certain duplicates
-	if err := r.db.CreateInBatches(params, batchSize).Error; err != nil {
-		// remove duplicate rows and try again
-		dedupedParams, errRemovingMatches := r.removeExactMatches(ctx, params)
-		if errRemovingMatches != nil {
-			return eris.Wrap(errRemovingMatches, "error removing duplicates in batch")
+	if err := r.db.Transaction(func(tx *gorm.DB) error {
```
Comment (dave-gantenbein) [inline pkg/api/mlflow/dao/repositories/param.go:60]: Is this just tested via the integration tests? Is it not possible to unit test this method?

File: pkg/api/mlflow/service/run/service.go [hunk 11]
```diff
@@ -520,7 +520,11 @@ func (s Service) LogParam(
 
 	param := convertors.ConvertLogParamRequestToDBModel(run.ID, req)
 	if err := s.paramRepository.CreateBatch(ctx, 1, []models.Param{*param}); err != nil {
-		return api.NewInternalError("unable to insert params for run '%s': %s", run.ID, err)
+		if _, ok := err.(repositories.ParamConflictError); ok {
```
Comment (dave-gantenbein) [inline pkg/api/mlflow/service/run/service.go:523]: It seems like it would be nice to have unit tests demonstrating the different error cases.

File: pkg/api/mlflow/dao/repositories/param.go (hunk 10 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:60]: I think not very useful to unit test since we are depending on database here

File: pkg/api/mlflow/service/run/service.go (hunk 11 above)
Comment (suprjinx) [reply pkg/api/mlflow/service/run/service.go:523]: I think we do cover this in service_test.go

Comment (suprjinx) [reply pkg/api/mlflow/service/run/service.go:523]: i can add same test for LogBatch..

Comment (dsuhinin) [inline pkg/api/mlflow/service/run/service.go:523]: probably we can just use `errors.Is` in that case?

**Review by dsuhinin — CHANGES_REQUESTED** [review]

File: pkg/api/mlflow/dao/repositories/param.go [hunk 12]
```diff
@@ -32,48 +57,68 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
+		if err := tx.WithContext(ctx).Clauses(clause.OnConflict{
+			Columns:   []clause.Column{{Name: "run_uuid"}, {Name: "key"}},
+			DoNothing: true,
+		}).CreateInBatches(params, batchSize).Error; err != nil {
+			return eris.Wrap(tx.Error, "error creating params in batch")
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:65]: yep!

File: pkg/api/mlflow/dao/repositories/param.go [hunk 13]
```diff
@@ -32,48 +57,68 @@ func NewParamRepository(db *gorm.DB) *ParamRepository {
...
-	for _, foundKey := range foundKeys {
-		delete(paramMap, foundKey)
+// makeSqlPlaceholders collects a string of (?,?,?), (?,?,?), etc, to the length of the params.
+func makeSqlPlaceholders(params []models.Param) string {
+	valuesArray := make([]string, len(params))
+	for i := range params {
+		valuesArray[i] = "(?,?,?)"
 	}
+	return strings.Join(valuesArray, ",")
+}
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:110]: I was trying to balance the "values" function which is highly type-specific, but this could just be `makePlaceholders(numberForEach int, total int) string` or the like

[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go:110]: or maybe return placeholders and values as a tuple from one function

File: pkg/api/mlflow/service/run/service.go [hunk 14]
```diff
@@ -607,7 +611,11 @@ func (s Service) LogBatch(
 		return api.NewInvalidParameterValueError(err.Error())
 	}
 	if err := s.paramRepository.CreateBatch(ctx, 100, params); err != nil {
-		return api.NewInternalError("unable to insert params for run '%s': %s", run.ID, err)
+		if _, ok := err.(repositories.ParamConflictError); ok {
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/service/run/service.go:614]: i tried and did not seem to work, but I'll look at the `errors` source

File: pkg/api/mlflow/dao/repositories/param.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply pkg/api/mlflow/dao/repositories/param.go]: I like this second idea!
a generic function that returns both the placeholders string and the values slice sounds like the way to go — once done, please change the other places in our code that use the same pattern (in a separate PR of course)

File: pkg/api/mlflow/dao/repositories/param.go [hunk 15]
```diff
@@ -3,13 +3,38 @@ package repositories
 import (
 	"context"
 	"fmt"
+	"strings"
 
 	"github.com/G-Research/fasttrackml/pkg/api/mlflow/dao/models"
 
 	"github.com/rotisserie/eris"
 	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go]: yep

File: pkg/api/mlflow/service/run/service.go [hunk 16]
```diff
@@ -607,7 +611,11 @@ func (s Service) LogBatch(
...
+			return api.NewInvalidParameterValueError("unable to insert params for run '%s': %s", run.ID, err)
+		} else {
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/service/run/service.go:616]: ok

File: pkg/api/mlflow/service/run/service.go [hunk 17]
```diff
@@ -520,7 +520,11 @@ func (s Service) LogParam(
 
 	param := convertors.ConvertLogParamRequestToDBModel(run.ID, req)
 	if err := s.paramRepository.CreateBatch(ctx, 1, []models.Param{*param}); err != nil {
-		return api.NewInternalError("unable to insert params for run '%s': %s", run.ID, err)
+		if _, ok := err.(repositories.ParamConflictError); ok {
+			return api.NewInvalidParameterValueError("unable to insert params for run '%s': %s", run.ID, err)
+		} else {
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/service/run/service.go]: ok

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 18]
```diff
@@ -111,9 +150,63 @@ func (s *LogParamTestSuite) Test_Error() {
...
+			http.MethodPost,
+		).WithRequest(
+			req,
+		).WithResponse(
+			&resp,
+		).DoRequest(
+			fmt.Sprintf("%s%s", mlflow.RunsRoutePrefix, mlflow.RunsLogParameterRoute),
+		),
+	)
+	//nolint:lll
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go]: done

File: pkg/api/mlflow/dao/repositories/param.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/param.go]: i don't think generic works very well here since we need specific fields from the struct, peculiar to a sql statement

File: pkg/api/mlflow/service/run/service.go (hunk 11 above)
Comment (suprjinx) [reply pkg/api/mlflow/service/run/service.go:523]: errors.As is working, changed to that

File: pkg/api/mlflow/service/run/service.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (suprjinx) [reply pkg/api/mlflow/service/run/service.go:613]: errors.As works

File: pkg/api/mlflow/dao/repositories/helpers.go [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+	valuesArray := make([]interface{}, len(params)*3)
+	index := 0
+	for _, param := range params {
+		valuesArray[index] = param.Key
+		valuesArray[index+1] = param.Value
+		valuesArray[index+2] = param.RunID
+		index = index + 3
+	}
+	return placeholders, valuesArray
+}
```
Comment (dsuhinin) [inline pkg/api/mlflow/dao/repositories/helpers.go:38]: we can move of course with this solution but my proposal was to have a function based on generics, where you can get the number of properties through the reflection and build `placeholders` for any slice. @jgiannuzzi what do you think?

File: pkg/api/mlflow/service/run/service.go [hunk 20]
```diff
@@ -520,6 +521,9 @@ func (s Service) LogParam(
 
 	param := convertors.ConvertLogParamRequestToDBModel(run.ID, req)
 	if err := s.paramRepository.CreateBatch(ctx, 1, []models.Param{*param}); err != nil {
+		if errors.As(err, &repositories.ParamConflictError{}) {
```
Comment (dsuhinin) [inline pkg/api/mlflow/service/run/service.go:524]: nice!

**Review by dsuhinin — DISMISSED** [review]
Looks good for me!

File: pkg/api/mlflow/service/run/service.go [hunk 21]
```diff
@@ -520,11 +521,10 @@ func (s Service) LogParam(
 
 	param := convertors.ConvertLogParamRequestToDBModel(run.ID, req)
 	if err := s.paramRepository.CreateBatch(ctx, 1, []models.Param{*param}); err != nil {
-		if _, ok := err.(repositories.ParamConflictError); ok {
+		if errors.As(err, &repositories.ParamConflictError{}) {
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/service/run/service.go:524]: should this not be `errors.Is` instead so we don't use the unwrapped error below and potentially lose parts of the error stack?

File: pkg/api/mlflow/service/run/service.go [hunk 22]
```diff
@@ -611,11 +611,10 @@ func (s Service) LogBatch(
 		return api.NewInvalidParameterValueError(err.Error())
 	}
 	if err := s.paramRepository.CreateBatch(ctx, 100, params); err != nil {
-		if _, ok := err.(repositories.ParamConflictError); ok {
+		if errors.As(err, &repositories.ParamConflictError{}) {
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/service/run/service.go:614]: same remark as above

File: pkg/api/mlflow/dao/repositories/helpers.go (hunk 19 above)
Comment (jgiannuzzi) [reply pkg/api/mlflow/dao/repositories/helpers.go:38]: agreed, I would have liked to see something more generic — however this is good enough for this PR

File: pkg/api/mlflow/dao/repositories/helpers.go [hunk 23]
```diff
@@ -0,0 +1,38 @@
+package repositories
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/G-Research/fasttrackml/pkg/api/mlflow/dao/models"
+)
+
+// makeSqlPlaceholders collects a string of (?,?,?), (?,?,?), etc
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/dao/repositories/helpers.go:10]: please provide a better description

File: pkg/api/mlflow/dao/repositories/helpers.go [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+	placeholderArray := make([]string, numberInEachSet)
+	for i := 0; i < numberInEachSet; i++ {
+		placeholderArray[i] = "?"
+	}
+	setsArray := make([]string, numberOfSets)
+	for i := 0; i < numberOfSets; i++ {
+		setsArray[i] = fmt.Sprintf("(%s)", strings.Join(placeholderArray, ","))
+	}
+	return strings.Join(setsArray, ",")
+}
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/dao/repositories/helpers.go:21]: this could be improved in terms of memory usage by using string builders (especially for the second iteration)

you may ask "is that not premature optimization?" and the answer to that is right up in the project name — we do care _a lot_ about optimizations in this project, and this is our main differentiator compared to the project whose API we implement

File: pkg/api/mlflow/dao/repositories/param.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply pkg/api/mlflow/dao/repositories/param.go]: the ordered list of fields would be part of the parameters to this generic function — anyway as I commented in another place, this is good enough for this PR

File: pkg/api/mlflow/dao/repositories/param.go (hunk 9 above)
Comment (jgiannuzzi) [reply pkg/api/mlflow/dao/repositories/param.go]: yes, it looks good now!

File: pkg/api/mlflow/dao/repositories/helpers.go (hunk 24 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/helpers.go:21]: Are not preallocated slices of the exact length efficient? The resulting arrays make it easier to use `strings.Join` to get the commas right.

Comment (jgiannuzzi) [reply pkg/api/mlflow/dao/repositories/helpers.go:21]: It doubles the allocated size: one for the slice of strings, and one for the joined string. It also puts more load on the GC by allocating many small strings (a slice of strings is just an array of pointers, and all of these pointer references need to be tracked by the GC)

Comment (jgiannuzzi) [reply pkg/api/mlflow/dao/repositories/helpers.go:21]: If you want to keep this implementation for readability, then you'd at least want to have the call to `fmt.Sprintf` done only once so that it's the same string reference that gets added n time to a slice

File: pkg/api/mlflow/service/run/service.go (hunk 21 above)
Comment (suprjinx) [reply pkg/api/mlflow/service/run/service.go:526]: I think we have to change from Error interface to error instance for `Is` to work.

Comment (suprjinx) [reply pkg/api/mlflow/service/run/service.go:526]: I'd suggest we keep the error interface and `As` usage so that each ParamConflictError can carry the particular message.

File: pkg/api/mlflow/dao/repositories/helpers.go (hunk 23 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/helpers.go:10]: added some description

File: pkg/api/mlflow/service/run/service.go (hunk 21 above)
Comment (jgiannuzzi) [reply pkg/api/mlflow/service/run/service.go:526]: I looked into this further, and we could make `Is` work by implementing `func (e ParamConflictError) Is(target error) bool`. However this does not seem necessary (at least at this stage), as indeed `As` does what we want here. I originally thought that `err` would get unwrapped and changed by the call to `As`, but this is not the case. It's the second parameter that gets assigned, but you don't use it so all is good!

File: pkg/api/mlflow/dao/repositories/helpers.go [hunk 25]
```diff
@@ -7,15 +7,17 @@ import (
...
 func makeSqlPlaceholders(numberInEachSet, numberOfSets int) string {
 	placeholderArray := make([]string, numberInEachSet)
 	for i := 0; i < numberInEachSet; i++ {
 		placeholderArray[i] = "?"
 	}
+	placeholderTemplate := strings.Join(placeholderArray, ",")
 	setsArray := make([]string, numberOfSets)
 	for i := 0; i < numberOfSets; i++ {
-		setsArray[i] = fmt.Sprintf("(%s)", strings.Join(placeholderArray, ","))
+		setsArray[i] = fmt.Sprintf("(%s)", placeholderTemplate)
```
Comment (jgiannuzzi) [inline pkg/api/mlflow/dao/repositories/helpers.go:20]: this call to `fmt.Sprintf` should be placed outside the loop

File: pkg/api/mlflow/dao/repositories/helpers.go (hunk 24 above)
Comment (jgiannuzzi) [reply pkg/api/mlflow/dao/repositories/helpers.go]: so I ran a quick benchmark of this implementation with the following code:
```go
func Benchmark_makeSqlPlaceholders(b *testing.B) {
	for i := 0; i < b.N; i++ {
		makeSqlPlaceholders(1, 100)
		makeSqlPlaceholders(2, 100)
		makeSqlPlaceholders(3, 100)
	}
}
```

I get the following results:
```
Benchmark_makeSqlPlaceholders-6            51420             22895 ns/op           13892 B/op        611 allocs/op
```

When I switch to using `strings.Builder` instead, I get the following results:
```
Benchmark_makeSqlPlaceholders-6           540442              2178 ns/op            3088 B/op          6 allocs/op
```
In this particular benchmark, it is 10x faster, uses 4x less memory and does 100x less allocations.

Here is the implementation:
```go
func makeSqlPlaceholders(numberInEachSet, numberOfSets int) string {
	setLength := 2*numberInEachSet + 1
	var set strings.Builder
	set.Grow(setLength)
	set.WriteString("(")
	for i := 0; i < numberInEachSet-1; i++ {
		set.WriteString("?")
		set.WriteString(",")
	}
	set.WriteString("?)")
	setString := set.String()

	setsLength := 2*numberOfSets*setLength - 1
	var sets strings.Builder
	sets.Grow(setsLength)
	for i := 0; i < numberOfSets-1; i++ {
		sets.WriteString(setString)
		sets.WriteString(",")
	}
	sets.WriteString(set.String())
	return sets.String()
}
```

I then tried another implementation that uses `strings.Repeat` and got these results:
```
Benchmark_makeSqlPlaceholders-6           927678              1131 ns/op            3985 B/op         15 allocs/op
```
It's another 2x faster and uses ~30% more memory and 2x more allocations, but it's much simpler:
```go
func makeSqlPlaceholders(numberInEachSet, numberOfSets int) string {
	set := fmt.Sprintf("(%s)", strings.Repeat("?,", numberInEachSet-1)+"?")
	return strings.Repeat(set+",", numberOfSets-1) + set
}
```

Again, I know this is not on the critical path, but we should always strive to have great performance anywhere in this project as this is our crucial differentiator!

Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/helpers.go]: okay, thanks for digging in -- I'll switch to the strings.Repeat implementation

File: pkg/api/mlflow/dao/repositories/helpers.go (hunk 25 above)
Comment (suprjinx) [reply pkg/api/mlflow/dao/repositories/helpers.go:20]: ok

**Review by jgiannuzzi — DISMISSED** [review]

File: tests/integration/golang/mlflow/run/log_batch_test.go [hunk 26]
```diff
@@ -123,6 +123,14 @@ func (s *LogBatchTestSuite) TestParams_Ok() {
 	})
 	require.Nil(s.T(), err)
 
+	// create preexisting param (other batch) for conflict testing
+	_, err = s.ParamFixtures.CreateParam(context.Background(), &models.Param{
+		RunID: run.ID,
+		Key:   "key1",
+		Value: "value1",
+	})
+	assert.Nil(s.T(), err)
```
Comment (jgiannuzzi) [inline tests/integration/golang/mlflow/run/log_batch_test.go:132]: ah shoot, we have merged #584 in the meantime, so you should use `require.Nil` instead of `assert.Nil` to be consistent!
sorry for the extra work 😞

Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_batch_test.go]: done

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 27]
```diff
@@ -107,9 +146,63 @@ func (s *LogParamTestSuite) Test_Error() {
...
+	assert.Empty(s.T(), resp)
+
+	// error conditions
+
+	// missing run_id
+	req = request.LogParamRequest{
+		Key:   "key1",
+		Value: "value1",
+	}
+	assert.Nil(
```
Comment (jgiannuzzi) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:158]: another one

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 28]
```diff
@@ -107,9 +146,63 @@ func (s *LogParamTestSuite) Test_Error() {
...
 		resp.Error(),
 	)
+
+	// conflicting param
+	req = request.LogParamRequest{
+		RunID: run.ID,
+		Key:   "key1",
+		Value: "value2",
+	}
+	assert.Nil(
```
Comment (jgiannuzzi) [inline tests/integration/golang/mlflow/run/log_parameter_test.go:182]: and the last one

File: tests/integration/golang/mlflow/run/log_parameter_test.go [hunk 29]
```diff
@@ -79,18 +79,57 @@ func (s *LogParamTestSuite) Test_Ok() {
...
 	)
 	assert.Empty(s.T(), resp)
+
+	// log duplicate, which is OK
+	req = request.LogParamRequest{
+		RunID: run.ID,
+		Key:   "key1",
+		Value: "value1",
+	}
+	assert.Nil(
```
[the remark this answers is not in the document]
Comment (suprjinx) [reply tests/integration/golang/mlflow/run/log_parameter_test.go]: done

**Review by jgiannuzzi — APPROVED** [review]
